File: python_vectorized/car.py
import numpy as np
from utils import *
from collections import deque
import time
import math

COLLIDE_REWARD = -2
GATE_REWARD = 1
PASS_REWARD = -.001

class Car:
    def __init__(self, x, y, width, height, rayCount=16, rayLength=200, maxSpeed=2):
        self.x = x
        self.y = y
        self.xI = x
        self.yI = y
        self.vx = 0
        self.vy = 0
        self.speed = 0
        self.maxSpeed = maxSpeed
        self.width = width
        self.height = height
        self.direction = 0
        self.rayCount = rayCount
        self.rayLength = rayLength
        self.lines = []
        self.gates = []
        self.gatesPassed = []
        self.numGatesPassed = 0
        self.lock = -1
        self.collide_threshold_sq = max(self.width, self.height)**2
        self.stackedStates = deque()
        self.stackedCount = 4
    
    def getInfo(self):
        return { 
            'x': self.x,
            'y': self.y,
            'direction': self.direction,
            'gatesPassed': self.gatesPassed.copy()
        }

    def reset(self):
        self.x = self.xI + np.random.uniform(-10, 10)
        self.y = self.yI + np.random.uniform(-10, 10)
        self.direction = 0
        self.gatesPassed = [False] * len(self.gates)
        self.lock = -1

    def setLines(self, lines):
        self.lines = np.array(lines)
    
    def setGates(self, gates):
        self.gates = np.array(gates)
        self.gatesPassed = [False] * len(gates)
    
    # lines is numpy array
    def getState(self):
        numLines = len(self.lines)
        rayCount = self.rayCount
        rayLength = self.rayLength
        maxSpeed = self.maxSpeed
        # each ray repeated numLines times
        origin = np.full((rayCount*numLines, 2), [self.x, self.y])
        theta = np.linspace(0, 2*np.pi, rayCount, endpoint=False)
        ray_dirs = np.array([np.cos(theta + self.direction), np.sin(theta + self.direction)]).T # rayCount x 2
        ray_dirs = np.repeat(ray_dirs, numLines, axis=0) # rayCount*numLines x 2
        p1 = np.tile(np.take(self.lines, 0, 1), (rayCount, 1))
        p2 = np.tile(np.take(self.lines, 1, 1), (rayCount, 1))
        # calculate dist to each line
        dists = ray_lineseg_intersection(origin, ray_dirs, p1, p2)
        dists = dists.reshape((rayCount, numLines))
        minI = np.argmin(dists, axis=1)
        minD = np.take_along_axis(dists, minI[None].T, 1).flatten()
        minD = np.where(minD > rayLength, rayLength, minD)
        vision = minD / rayLength
        feat = [self.speed/maxSpeed]
        combined = np.append(vision, feat)
        if len(self.stackedStates) == 0:
            for _ in range(self.stackedCount):
                self.stackedStates.append(combined)
        else:
            self.stackedStates.popleft()
            self.stackedStates.append(combined)
        return np.concatenate(self.stackedStates)
    
    def step(self, action):
        if action == 0:
            self.speed = min(self.speed + 0.2, self.maxSpeed)
        elif action == 1:
            self.speed = max(self.speed - 0.2, -self.maxSpeed/2)
        if action == 2:
            self.direction -= 0.04
        elif action == 3:
            self.direction += 0.04
        evt = self.update()
        reward = 0
        terminated = False
        if evt == 'collided':
            reward = COLLIDE_REWARD
            terminated = True
        elif evt == 'allgates':
            reward = GATE_REWARD
            terminated = True
        elif evt == 'gate':
            reward = GATE_REWARD
        else:
            reward = 0 # np.mean(self.getState()[:-1]) * .1
        return reward, terminated

    
    def update(self):
        c = np.cos(self.direction)
        s = np.sin(self.direction)
        self.vx = self.speed * c
        self.vy = self.speed * s
        self.x += self.vx
        self.y += self.vy
        self.speed *= 0.99

        numCarPoints = 4
        carPoints = np.array([[-self.width/2, self.height/2], 
                        [self.width/2, self.height/2],
                        [self.width/2, -self.height/2],
                        [-self.width/2, -self.height/2]])        
        rot = np.array([[c, s], [-s, c]])
        carPoints = np.matmul(carPoints, rot)
        carPoints += np.array([self.x, self.y])

        numLines = len(self.lines)
        centers = np.full((numLines, 2), [self.x, self.y])
        p1 = np.take(self.lines, 0, 1)
        p2 = np.take(self.lines, 1, 1)
        dists = min_distance_sq(centers, p1, p2)
        lines = self.lines[dists < self.collide_threshold_sq]
        numLines = len(lines)

        if numLines > 0:
            carPointsRep = np.repeat(carPoints, numLines, axis=0)
            carPointsRepShifted = np.roll(carPointsRep, numLines, axis=0)
            p1 = np.tile(np.take(lines, 0, 1), (numCarPoints, 1))
            p2 = np.tile(np.take(lines, 1, 1), (numCarPoints, 1))

            lineIntersections = linesegs_intersect(carPointsRep, carPointsRepShifted, p1, p2)
            if np.any(lineIntersections):
                return 'collided'

        numGates = len(self.gates)
        carPointsRep = np.repeat(carPoints, numGates, axis=0)
        carPointsRepShifted = np.roll(carPointsRep, numGates, axis=0)
        newGatePassed = False
        for j in range(len(self.gates)):
            if self.lock == j:
                continue
            g = self.gates[j]
            p1 = g[0][np.newaxis]
            p2 = g[1][np.newaxis]
            if not self.gatesPassed[j] and np.any(linesegs_intersect(carPointsRep, carPointsRepShifted, p1, p2)):
                self.gatesPassed[j] = True
                newGatePassed = True
                self.lock = j
                self.numGatesPassed += 1
                if all(self.gatesPassed):
                    self.gatesPassed = [False] * len(self.gates)
                    return 'allgates'
        if newGatePassed:
            return 'gate'
        return 'none'

File: js_tensorflow/car.js

const maxSpeed = 2
let carImg

class Car {    

    constructor(x, y, width, height) {
        this.x = x
        this.y = y
        this.xI = x
        this.yI = y
        this.vx = 0
        this.vy = 0
        this.speed = 0
        this.width = width
        this.height = height
        this.direction = 0
        this.gatesPassed = []
        this.gates = []
        this.lock = -1
    }

    reset() {
        this.x = this.xI
        this.y = this.yI
        this.direction = 0
        this.gatesPassed = Array(gates.length).fill(false)
        this.lock = -1
    }

    getState(lines, drawEnabled) {
        const rayLength = 200
        let vision = []
        for(let theta = 0; theta < Math.PI*2; theta += Math.PI/6) {
            let x2 = this.x + rayLength * Math.cos(theta + this.direction)
            let y2 = this.y + rayLength * Math.sin(theta + this.direction)
            x2 = Math.round(x2 * 10) / 10
            y2 = Math.round(y2 * 10) / 10
            let closest = [x2, y2]
            let minDistSq = rayLength*rayLength
            for(let l of lines) {
                let cross = intersectsWhere(this.x, this.y, x2, y2, l.x1, l.y1, l.x2, l.y2)
                let distSq = (cross[0] - this.x)*(cross[0] - this.x) + (cross[1] - this.y)*(cross[1] - this.y)
                if(cross && distSq < minDistSq) {
                    closest = cross 
                    minDistSq = distSq
                }
            }
            if(drawEnabled) {
                stroke(255)
                strokeWeight(1)
                line(this.x, this.y, closest[0], closest[1])
                strokeWeight(4)
                point(closest[0], closest[1])
            }
            vision.push(Math.sqrt(minDistSq) / rayLength)
        }
        vision.push(this.speed / maxSpeed) // add speed to feature list
        return vision
    }

    setGates(gates) {
        this.gates = gates
        this.gatesPassed = Array(gates.length).fill(false)
    }

    forward() {
        //this.vx = Math.min(this.vx + .1 * Math.cos(this.direction), Math.sqrt(2))
        //this.vy = Math.min(this.vy + .1 * Math.sin(this.direction), Math.sqrt(2))
        //this.x += Math.cos(this.direction) * 4
        //this.y += Math.sin(this.direction) * 4
        this.speed = Math.min(this.speed + .2, maxSpeed)
    }
    backward() {
        //this.vx = Math.max(this.vx - .1 * Math.cos(this.direction), -Math.sqrt(2))
        //this.vy = Math.max(this.vy - .1 * Math.sin(this.direction), -Math.sqrt(2))
        //this.x -= Math.cos(this.direction) * 4
        //this.y -= Math.sin(this.direction) * 4
        this.speed = Math.max(this.speed - .2, -maxSpeed)
    }
    left() {
        this.direction -= .04
    }
    right() {
        this.direction += .04
    }

    update(lines, drawEnabled) {
        this.vx = this.speed * Math.cos(this.direction)
        this.vy = this.speed * Math.sin(this.direction)
        this.x += this.vx
        this.y += this.vy
        this.speed *= 0.99

        // "top left" => "cw"
        let c = Math.cos(this.direction)
        let s = Math.sin(this.direction)
        const carPoints = [ 
            [-this.width/2, this.height/2],
            [this.width/2, this.height/2],
            [this.width/2, -this.height/2],
            [-this.width/2, -this.height/2]
        ]
        for(let point of carPoints) {
            let x1 = point[0]
            let y1 = point[1]
            point[0] = c*x1 - s*y1
            point[1] = s*x1 + c*y1
        }
        let newGatePassed = false
        for(let i = 0; i < carPoints.length; i++) {
            let p1 = carPoints[i]
            let p2 = carPoints[(i + 1) % carPoints.length]
            //stroke(255, 0, 0)
            //strokeWeight(2)
            //line(p1[0] + this.x, p1[1] + this.y, p2[0] + this.x, p2[1] + this.y)
            for(let l of lines) {
                if(intersects(p1[0] + this.x, p1[1] + this.y, p2[0] + this.x, p2[1] + this.y, l.x1, l.y1, l.x2, l.y2)) {
                    if(drawEnabled) {
                        stroke(255, 0, 0)
                        strokeWeight(4)
                        line(l.x1, l.y1, l.x2, l.y2)
                    }
                    return 'collided'
                }
            }
            for(let j = 0; j < this.gates.length; j++) {
                if(this.lock == j) continue
                let g = this.gates[j]
                if(!this.gatesPassed[j] && intersects(p1[0] + this.x, p1[1] + this.y, p2[0] + this.x, p2[1] + this.y, g.x1, g.y1, g.x2, g.y2)) {
                    this.gatesPassed[j] = true
                    newGatePassed = true
                    this.lock = j
                    if(this.gatesPassed.every(x => x))
                        this.gatesPassed = Array(gates.length).fill(false)
                }
            }
        }
        if(newGatePassed) return 'gate'
        return 'none'
    }

    draw() {
        fill(255, 127, 0)
        stroke(0)
        strokeWeight(1)
        push()
        translate(this.x, this.y)
        rotate(this.direction)
        image(carImg, -this.width/2, -this.height/2, this.width, this.height)
        pop()
    }

    static load() {
        carImg = loadImage('car_small.png')
    }

}